code/tests/generic_002.cpp

```diff
@@ -0,0 +1,176 @@
+  }
+
+  virtual int run() override {
+  
+    const int fd_foo = open(foo_path.c_str(), O_RDWR | O_CREAT, TEST_FILE_PERMS);
+    if (fd_foo < 0) {
+      return -1;
+    }
+
+    // Create new set of links
+    for (int i = 0; i < NUM_LINKS; i++){
+      string foo_link = foo_link_path + std::to_string(i);
+      if(link(foo_path.c_str(), foo_link.c_str()) < 0){
+        return -2;
+      }
+
+      // fsync foo
+      int res = fsync(fd_foo);
+      if (res < 0){
+        return -3;
+      }
+
+      // Make a user checkpoint here
+      if (Checkpoint() < 0){
+        return -4;
+      }
+    }
+
+    // Remove the set of added links
+    for (int i = 0; i < NUM_LINKS; i++){
+      string foo_link = foo_link_path + std::to_string(i);
+      if(remove(foo_link.c_str()) < 0) {
+        return -5;
+      }
+
+      // fsync foo
+      int res = fsync(fd_foo);
+      if (res < 0){
+        return -6;
+      }
+
+      // Make a user checkpoint here
+      if (Checkpoint() < 0){
+        return -7;
+      }
+    }
+
+    //Close open files  
+    close(fd_foo);
+
+    return 0;
+  }
+
+  virtual int check_test(unsigned int last_checkpoint,
+      DataTestResult *test_result) override {
+
+    char mode[] = "0777";
+    int i_mode = strtol(mode,0,8);
+    chmod(foo_path.c_str(), i_mode);
+
+    // Since crash could have happened between a fsync and Checkpoint, last_checkpoint only tells the minimum number of operations that have
+    // happened. In reality, one more fsync could have happened, which might alter the link count, delta is used to keep track of that. For example,
+    // during linking phase, if crash occured between fsync and checkpoint, number of links will be 1 more than what checkpoint says while during remove
+    // phase, the same will be 1 less than what the checkpoint says.
+    int delta = 0;
+    if (last_checkpoint == 2 * NUM_LINKS) { // crash after all fsyncs completed
+    	delta = 0;
+    } else if (last_checkpoint < NUM_LINKS) { // crash during link phase
+    	delta = 1;
+    } else { // crash during remove phase
+    	delta = -1;
+    }
+
+    int num_expected_links;
+    if (last_checkpoint <= NUM_LINKS) {
+    	num_expected_links = last_checkpoint;
+    } else {
+    	num_expected_links = 2 * NUM_LINKS - last_checkpoint;
+    }
+
+    struct stat st;
+    stat(foo_path.c_str(), &st);
+    int actual_count = st.st_nlink;
+    actual_count--; // subtracting one for the file itself
+
+    // If actual count doesn't match with expected number of links
+    if (actual_count != num_expected_links && actual_count != num_expected_links + delta) {
+        test_result->SetError(DataTestResult::kFileMetadataCorrupted);
+        test_result->error_description = " : Number of links not matching with expected count";
+    }
+
+    return 0;
+  }
+
+   private:
+    const string foo_path = TEST_MNT "/" TEST_DIR_A "/" TEST_FILE_FOO;    
+    const string foo_link_path = TEST_MNT "/" TEST_DIR_A "/" TEST_FILE_FOO_LINK;
+    
```

Comment:
Modify these definitions to use mnt_dir_ which is initialized in base_test case. When we parallelize crashmonkey runs, this would be required. Refer to this test file https://github.com/utsaslab/crashmonkey/blob/pandian-port-xfstests/code/tests/btrfs_rename_symlink.cpp and make changes accordingly.

Reply:
Sure. I see many other test files which do not use mnt_dir_ but use TEST_MNT as a hardcoded macro (like https://github.com/utsaslab/crashmonkey/blob/pandian-port-xfstests/code/tests/generic_066.cpp, https://github.com/utsaslab/crashmonkey/blob/pandian-port-xfstests/code/tests/generic_059.cpp, https://github.com/utsaslab/crashmonkey/blob/pandian-port-xfstests/code/tests/generic_041.cpp, https://github.com/utsaslab/crashmonkey/blob/pandian-port-xfstests/code/tests/generic_034.cpp). Those should also be modified to use mnt_dir_ ?

Reply:
Yes, they are the tests created before we discussed about parallelizing CM. We will need to clean them up when we merge parallelism code. For now, let's avoid the hardcoding in any future standalone tests.

Reply:
Cool.